### ISSUE TITLE
Redux store

### ISSUE THREAD

**Issue by flaviocarvalho** [rt2zz/redux-persist#419]
So I'm trying to implement v5 on a react-native project, I've never used any version of this library before.
This is what I have right now and the issue I'm having is my `mapStateToProps` state is returning:
`{ registry: ['root'], bootstrapped: false }`
I think this might be related to https://github.com/rt2zz/redux-persist/issues/408, let me know what you guys think

```
redux-persist@5.0.0-beta.1
```

`store.js`
```import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { persistStore } from 'redux-persist'

import { check } from 'src/utils'
import reducers from 'src/reducers'

const reduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

const applyReduxDevTools = check.isDevAndDebugging && reduxDevTools

const store = createStore(
  reducers,
  applyReduxDevTools
    ? compose(applyMiddleware(thunk), applyReduxDevTools)
    : compose(applyMiddleware(thunk))
)

export default persistStore(store)
```

`reducers.js`
```
import { combineReducers } from 'redux'
import { persistReducer } from 'redux-persist'
import { AsyncStorage } from 'react-native'

import x from 'src/reducers/x'
import y from 'src/reducers/y'

const config = {
  version: '0.0.1',
  key: 'root',
  storage: AsyncStorage,
  debug: true
}

const rootReducer = combineReducers({ x, y })

export default persistReducer(config, rootReducer)
```

**Comment by rt2zz** [comment]
right so the issue is in store.js. the fix:
```js
import thunk from 'redux-thunk'
import { persistStore } from 'redux-persist'

import { check } from 'src/utils'
import reducers from 'src/reducers'

const reduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

const applyReduxDevTools = check.isDevAndDebugging && reduxDevTools

const store = createStore(
  reducers,
  applyReduxDevTools
    ? compose(applyMiddleware(thunk), applyReduxDevTools)
    : compose(applyMiddleware(thunk))
)

const persistor = persistStore(store)
export default { store, persistor }
```

i.e. persistStore returns a persistor not the original store. I recommend you export both so you can pass store to `Provider` and pass persistor to `PersistGate` (a component to delay rendering until the rehydration is complete).

LMK If you have suggestions for how to improve the docs to avoid this confusion in the future.

**Comment by flaviocarvalho** [comment]
Awesome, I think that worked, I'm getting some import errors now, but I'm pretty sure it's unrelated to redux-persist.
Just more information about persistor vs store would be great, also a complete demo example, so people can follow the setup flow.

Thanks @rt2zz

**Comment by flaviocarvalho** [comment]
Didn't want to open a new issue, so I did what you said, and I also looked in the docs more:

```js
// store.js - (Btw you should add the async keyword in the docs)
export default async function configureStore () {
  const initState = await getStoredState(reducerConfig)

  const store = createStore(
    reducers,
    initState,
    check.isDevAndDebugging
      ? compose(applyMiddleware(thunk), check.isDevAndDebugging)
      : applyMiddleware(thunk)
  )

  listeners(store)

  const persistor = persistStore(store)

  return { store, persistor }
}
```
then in my entry point:

```js
import configureStore from 'src/store'

const { store, persistor } = configureStore()

console.log(configureStore(), 'configureStore')
// One of those have the store and persistor objects I'm looking for
// Promise {_40: 0, _65: 0, _55: null, _72: null}

console.log(store, 'store')
// undefined

const App = () => (
  <Provider store={store}>
    <PersistGate
      persistor={persistor}
      loading={<Loading />}
      onBeforeLift={() => console.log('lifted')}
    >
      <Container>
        <StatusBar hidden />

        <ScreenRoot />
      </Container>
    </PersistGate>
  </Provider>
)

export default App
```

Am I supposed to handle this promise with a class component and react cycles? I feel that it would be a hacky fix, thoughts on this @rt2zz ?

**Comment by rt2zz** [comment]
no need to use getStoredState. I would update configureStore as follows:
```js
// store.js - (Btw you should add the async keyword in the docs)
export default function configureStore () {
  const store = createStore(
    reducers,
    undefined,
    check.isDevAndDebugging
      ? compose(applyMiddleware(thunk), check.isDevAndDebugging)
      : applyMiddleware(thunk)
  )
   
  // @NOTE not sure what this does, assuming listeners is defined outside of this method?
  listeners(store)

  const persistor = persistStore(store)

  return { store, persistor }
}
```

**Comment by rt2zz** [comment]
also note: the reason you had a promise was that async functions always return promises. Make sure you remove the async flag on configureStore

**Comment by flaviocarvalho** [comment]
Got it, I thought I had to add that initial state, also to clarify I do need to update my redux store manually if a local version has been found right? I had to dispatch this in order for the PersistGate component to show the root component instead of the loading one:
```js
persistor.dispatch({ type: REHYDRATE })
```
Problem I'm having now is all my other actions are not working as before, maybe something to do with the async initial state I was doing? need to try it tomorrow.

**Comment by rt2zz** [comment]
no you should not need to manually do `persistor.dispatch({ type: REHYDRATE })` that happens under the hood. I think maybe you still need to wrap your reducer in persistReducer, e.g.
```js
import { persistStore, persistReducer } from 'redux-persist'
//...

export default function configureStore () {
  const persistedReducer = persistReducer(reducers)
  const store = createStore(
    persistedReducer,
    undefined,
    check.isDevAndDebugging
      ? compose(applyMiddleware(thunk), check.isDevAndDebugging)
      : applyMiddleware(thunk)
  )
   
  // @NOTE not sure what this does, assuming listeners is defined outside of this method?
  listeners(store)

  const persistor = persistStore(store)

  return { store, persistor }
}
```

**Comment by flaviocarvalho** [comment]
I'm already doing that:
```js
// src/reducers.js
export const reducerConfig = {
  version: VERSION,
  key: 'root',
  storage: AsyncStorage,
  debug: check.isDevAndDebugging
}

const rootReducer = combineReducers({
  a,
  b
})

export default persistReducer(reducerConfig, rootReducer)
```

**Comment by rt2zz** [comment]
:/ my bad, there was a bug beta.1, just released **@5.0.0-beta.2** which I have confirmed resolves the not persisting bug.

**Comment by flaviocarvalho** [comment]
Awesome, it works now, thank you so much @rt2zz :)
